### PR TITLE
add retry for TransferSui transaction builder call at faucet

### PR DESCRIPTION
**Pull request by longbowlu**
This PR adds a simple retry functionality for calling gateway's transaction building api `transfer_sui`. Also it extends the `RpcError` to include the error code for more info. 

## Context
We started to observe a very strange thing recently:
When running the cluster test  on the CI pipeline (aka the runner) against continuous cluster, gateway always fails `get_object` through aggregators, thus faucet errors out when building Transaction for `transferSui`. 

example error msg:
```
2022-08-12T21:06:29.352719Z ERROR sui_faucet::faucet::simple_faucet: Encountered error in transfer sui: Err(RPC call failed: ErrorObject { code: ServerError(-32000), message: "Too many authority errors were detected for get_object_by_id: [(k#839e99f8b03f0f5563d6cd9cc39e10a8cf483ad2775aecbb927031370925ef4e, RpcError(\"transport error\")), (k#89cafd721eeb1b13483d5fb0968500e78103e8146b45edb30d93781d946d0aeb, RpcError(\"transport error\")), (k#e97638a9e10b9cc46d85b81191a60a6a6fcae63dacd811c07502db9c5cfc55b5, RpcError(\"transport error\"))]", data: None }
```

However when triggering the test locally, it does not have this problem at all.. The interesting part is regardless of where the test harness runs, gateway shouldn't behave differently, not to mention faucet is hit in the same way too. 

We don't see this issue in any other apis such as `execution_transaction`, but probably just because they have retries already

### CODE REVIEW

**Comment by oxade** [comment]
@longbowlu great observation. If I understand correctly, we don't know the RC, but we this as a stopgap?

**Comment by longbowlu** [comment]
> @longbowlu great observation. If I understand correctly, we don't know the RC, but we this as a stopgap?

correct. I'm very confused. But gateway is really doing strange things recently.

**Comment by oxade** [comment]
> > @longbowlu great observation. If I understand correctly, we don't know the RC, but we this as a stopgap?
> 
> correct. I'm very confused. But gateway is really doing strange things recently.

good thing its going away :)

File: crates/sui-faucet/src/faucet/simple_faucet.rs [hunk 1]
```diff
@@ -178,21 +180,74 @@ impl SimpleFaucet {
...
         signer: SuiAddress,
         recipient: SuiAddress,
         budget: u64,
         amount: u64,
         uuid: Uuid,
-    ) -> Result<SuiTransactionResponse, anyhow::Error> {
-        let context = &self.wallet;
+    ) -> Result<TransactionData, anyhow::Error> {
+        // try 3 times in total
+        let retry_intervals_ms = [Duration::from_millis(500), Duration::from_millis(1000)];
```
Comment (oxade) [inline crates/sui-faucet/src/faucet/simple_faucet.rs:193]: nit: I would use constants or a counter here so its easy to see that it retries 3 times.

File: crates/sui-faucet/src/faucet/simple_faucet.rs [hunk 2]
```diff
@@ -178,21 +180,74 @@ impl SimpleFaucet {
...
+            .await;
+        let mut iter = retry_intervals_ms.iter();
+        while data.is_err() {
+            if let Some(duration) = iter.next() {
+                tokio::time::sleep(*duration).await;
+                debug!(?uuid, "Retrying constructing TransferSui txn");
+            } else {
+                warn!(
+                    ?uuid,
+                    "Failed to construct TransferSui txn after attempts: {:?}", &retry_intervals_ms
```
Comment (oxade) [inline crates/sui-faucet/src/faucet/simple_faucet.rs:205]: This measures time not # attempts, so the text is misleading.

**Review by oxade — CHANGES_REQUESTED** [review]

File: crates/sui-faucet/src/faucet/simple_faucet.rs (hunk 1 above)
Comment (longbowlu) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:193]: with a counter it's hard to see the intervals. I'd prefer to keep it in this way and change the comment to "retry 2 times"

**Review by oxade — APPROVED** [review]

**Pull request merged**